ananse/network.py

```diff
@@ -368,54 +373,84 @@ def aggregate_binding(
-        tmp = ddf.merge(gene_df, left_on="enhancer", right_index=True)
-
-        # Remove everything with weight 0
-        tmp = tmp[tmp["weight"] > 0]
+        # TODO: This is hacky (depending on "_"), however the hdf.keys() method is
+        # much slower.
+        all_tfs = [x for x in dir(hdf.root) if not x.startswith("_")]
+        logger.info(f"Binding file contains {len(all_tfs)} TFs.")
+        if tfs is None:
+            tfs = all_tfs
+        else:
+            tfs = [tf for tf in tfs if tf in all_tfs]
+            logger.info(f"Using {len(tfs)} TFs.")
+
+        # Read enhancer index from hdf5 file
+        enhancers = hdf.get(key="_index")
+        chroms = enhancers.index.to_series().str.replace(":.*", "").unique()
+
+        tmpdir = mkdtemp()
+        self._tmp_files.append(tmpdir)  # mark for deletion later
+
+        for chrom in chroms:
+            logger.info(f"Reading and aggregating binding from {chrom}")
+            idx = enhancers.index.str.contains(f"^{chrom}:")
+            idx_i = np.arange(enhancers.shape[0])[idx]
+
+            # Create a pyranges object
+            enhancer_pr = pr.PyRanges(
+                enhancers[idx]
+                .index.to_series()
+                .str.split(r"[:-]", expand=True)
+                .rename(columns={0: "Chromosome", 1: "Start", 2: "End"})
+            )
 
-        # Modify the binding by the weight, which is based on distance to TSS
-        tmp["weighted_binding"] = tmp["weight"] * tmp["binding"]
+            # Link enhancers to genes on basis of distance to annotated TSS
+            gene_df = self.enhancer2gene(
+                enhancer_pr,
+                up=up,
+                down=down,
+                alpha=alpha,
+                promoter=promoter,
+                full_weight_region=full_weight_region,
+            )
 
-        logger.info("Grouping by tf and target gene...")
-        # Using one column that combines TF and target as dask cannot handle MultiIndex
-        tmp["tf_target"] = tmp["factor"] + "_" + tmp["gene"]
+            gene_df = gene_df.dropna()
+
+            bp = pd.DataFrame(index=enhancers[idx].index)
+
+            for tf in tqdm(tfs):
+                # Load TF binding data for this chromosome.
+                # hdf.get() is *much* faster here than pd.read_hdf()
+                bp[tf] = hdf.get(key=tf)[idx_i].values
+
+            gene_df = gene_df[gene_df["weight"] > 0]
+
+            gene_df = gene_df.join(bp).dropna()
+            bp = gene_df[tfs]
+            gene_df = gene_df[["gene", "weight"]]
+            bp = bp.mul(gene_df["weight"], axis=0)
+            bp["gene"] = gene_df["gene"]
+            tmp = bp.groupby("gene")
+            if combine_function == "mean":
+                tmp = tmp.mean()
+            elif combine_function == "max":
+                tmp = tmp.max()
+            elif combine_function == "sum":
+                tmp = tmp.sum()
+            tmp = tmp.reset_index().melt(
```

Comment:
```
else:
    raise NotImplementedError
```

Reply:
It's already checked earlier in the function.